cmd/rpc.go

```diff
@@ -0,0 +1,337 @@
+	r.cmd.Flag("batch-sleep", "Sleep time between batches").IntVar(&r.batchSleep)
+	// TODO: implement me
+	//r.cmd.Flag("limit-seed", "Seed value for deterministic random batching").StringVar(&r.limitSeed)
+	// TODO: implement me
+	//r.cmd.Flag("limit", "Send request to only a subset of nodes, can be a percentage").StringVar(&r.limit)
+	// json is default and only means of output at this time (and perhaps forever)
+	//r.cmd.Flag("json", "Produce JSON output").Short('j').BoolVar(&r.json)
+	r.cmd.Flag("display", "Influence how results are displayed. One of ok, all or failed").StringVar(&r.display)
+	r.cmd.Flag("verbose", "Be verbose").Short('v').BoolVar(&r.verbose)
+
+	r.cmd.Flag("target", "Target messages to a specific sub collective").Short('T').StringVar(&r.target)
+	r.cmd.Flag("discovery-timeout", "Timeout for doing discovery").IntVar(&r.discoveryTimeout)
+	r.cmd.Flag("timeout", "Timeout for calling remote agents").Short('t').IntVar(&r.timeout)
+	r.cmd.Flag("quiet", "Do not be verbose").Short('q').BoolVar(&r.quiet)
+	r.cmd.Flag("ttl", "Set the message validity period").IntVar(&r.ttl)
+	r.cmd.Flag("reply-to", "Set a custom target for replies").StringVar(&r.replyTo)
+	r.cmd.Flag("disc-method", "Which discovery method to use").StringVar(&r.discMethod)
+	r.cmd.Flag("disc-option", "Options to pass to the discovery method").StringsVar(&r.discOptions)
+	// TODO: implement me
+	//r.cmd.Flag("nodes", "File with list of nodes to address").PlaceHolder("FILE").StringVar(&r.nodes)
+	r.cmd.Flag("publish_timeout", "Timeout for publishing requests to remote agents").IntVar(&r.publishTimeout)
+	r.cmd.Flag("threaded", "Start publishing requests and receiving responses in threaded mode").BoolVar(&r.threaded)
+	// TODO: implement me
+	//r.cmd.Flag("sort", "Sort the output of an RPC call before processing").BoolVar(&r.sort)
+	r.cmd.Flag("connection-timeout", "Set the timeout for establishing a connection to the middleware").IntVar(&r.connectionTimeout)
+
+	r.cmd.Flag("with", "Combined classes and facts filter").Short('W').StringsVar(&r.with)
+	// TODO: implement me correctly
+	r.cmd.Flag("select", "Compound filter combining facts and classes").Short('S').StringsVar(&r.sel)
+	r.cmd.Flag("with-fact", "Match hosts with a certain fact").Short('F').StringsVar(&r.withFact)
+	r.cmd.Flag("with-class", "Match hosts with a certain config management class").Short('C').StringsVar(&r.withClass)
+	r.cmd.Flag("with-agent", "Match hosts with a certain agent").Short('A').StringsVar(&r.withAgent)
+	r.cmd.Flag("with-identity", "Match hosts with a certain configured identity").Short('I').StringsVar(&r.withIdentity)
+
+	r.cmd.Arg("agent", "Agent to call").Required().StringVar(&r.agent)
+	r.cmd.Arg("action", "Action to call").Required().StringVar(&r.action)
+	r.cmd.Arg("key=val", "Arguments to pass to agent").StringMapVar(&r.args)
+
+	return nil
+}
+
+func (r *rpcCommand) Configure() error {
+
+	err = commonConfigure()
+	if err != nil {
+		return err
+	}
+
+	if r.threaded {
```

Comment:
I think a lot of these if statements here can be simplified by either having the cli options set the options directly and by letting the cli parsing default to whats in the config

Reply:
I do something like this which simplify filter stuff a lot:

```
var (
	factF     []string
	agentsF   []string
	classF    []string
	identityF []string
	combinedF []string
)

func addRPCFilter(cmd *kingpin.CmdClause) {
	cmd.Flag("wf", "Match hosts with a certain fact").Short('F').StringsVar(&factF)
	cmd.Flag("wc", "Match hosts with a certain configuration management class").Short('C').StringsVar(&classF)
	cmd.Flag("wa", "Match hosts with a certain Choria agent").Short('A').StringsVar(&agentsF)
	cmd.Flag("wi", "Match hosts with a certain Choria identity").Short('I').StringsVar(&identityF)
	cmd.Flag("with", "Combined classes and facts filter").Short('W').PlaceHolder("FILTER").StringsVar(&combinedF)
	cmd.Flag("batch", "Do requests in batches").Default("0").IntVar(&batchSize)
	cmd.Flag("batch-sleep", "Sleep time between batches").PlaceHolder("SECONDS").Default("10").IntVar(&batchSleep)
	cmd.Flag("nodes", "Text file with nodes to use for discovered noded").PlaceHolder("FILE").ExistingFileVar(&nodesFile)
}

func rpcFilter(agent string) (*protocol.Filter, error) {
	return client.NewFilter(
		client.FactFilter(factF...),
		client.AgentFilter(agentsF...),
		client.ClassFilter(classF...),
		client.IdentityFilter(identityF...),
		client.CombinedFilter(combinedF...),
	)
}
```

Reply:
I get segfaults when trying to simplify the if statements. For example:

```
r.cmd.Flag("target", "Target messages to a specific sub collective").Short('T').Default(cfg.MainCollective).StringVar(&cfg.MainCollective)
```

gets me this error when I run it:

```
[vagrant@puppet ~]$ choria rpc -q process list pattern=ruby user=root
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0xba16aa]

goroutine 1 [running]:
github.com/choria-io/go-choria/cmd.(*rpcCommand).Setup(0xc000001c80, 0x0, 0x0)
	/go/src/github.com/choria-io/go-choria/cmd/rpc.go:84 +0x77a
github.com/choria-io/go-choria/cmd.ParseCLI(0xc000155f78, 0xbc14dd)
	/go/src/github.com/choria-io/go-choria/cmd/cmd.go:53 +0x533
main.main()
	/go/src/github.com/choria-io/go-choria/main.go:15 +0x26
```

Reply:
ah i bet cfg isnt set yet when the cli arguments are parsed, lame :(

Reply:
Also, I can't seem to find the FactFilter, AgentFilter, ClassFilter, etc functions, though I was looking in https://github.com/choria-io/go-protocol/blob/master/protocol/filter.go

Reply:
you generally wouldnt need to muck about with protocol at all, there are 2 types of client.

https://github.com/choria-io/go-client and the rpc one you already know.  It's a bit like a TCP packet vs a HTTP packet, RPC requests are wrapped within normal client requests (they just move bytes).

Like in TCP addressing is in the underlying client, so the go-client have various helpers and discovery stuff etc. So in the example `client.NewFilter()` comes from this lower level client at https://github.com/choria-io/go-client/blob/master/client/filter.go

Reply:
Ok, thanks! I've simplified the filter configuration. Should I put back those if statements in `Configure` because of the segfaults?

Reply:
yeah i guess, I'll take a look at how i can make it easier when i look again after my trip

Reply:
corrections pushed!

Reply:
Regarding the segfaults when trying to simplify Configured() - you can do the `commonConfigure()` stuff in `Setup` prior to creating all the app stuff then adjust your `Configure()` to not call `commonConfigured()` again, that should solve it and simplify things a lot